pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -0,0 +1,867 @@
+						// update "ResolvedRefs" status true with "InvalidCertificateRef" reason
+						listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
+							Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+							Status:             metav1.ConditionFalse,
+							LastTransitionTime: metav1.Now(),
+							Reason:             string(v1alpha1.ListenerReasonInvalidCertificateRef),
+							Message:            msg,
+						})
+
+						logger.Errorf(msg)
+						listenerInError = true
+						continue
+					}
+
+					tlsConfigs[configKey] = tlsConf
+				}
+			}
+
+			// Supported Protocol
+			if listener.Protocol != v1alpha1.HTTPProtocolType && listener.Protocol != v1alpha1.HTTPSProtocolType {
+				msg := fmt.Sprintf("Unsupported listener protocol %q", listener.Protocol)
+				// update "Detached" status true with "UnsupportedProtocol" reason
+				listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
+					Type:               string(v1alpha1.ListenerConditionDetached),
+					Status:             metav1.ConditionTrue,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(v1alpha1.ListenerReasonUnsupportedProtocol),
+					Message:            msg,
+				})
+
+				logger.Errorf(msg)
+				listenerInError = true
+				continue
+			}
+
+			// Supported Route types
+			if listener.Routes.Kind != "HTTPRoute" {
+				msg := fmt.Sprintf("Unsupported Route Kind %q", listener.Routes.Kind)
+				// update "ResolvedRefs" status true with "InvalidRoutesRef" reason
+				listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
+					Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionFalse,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(v1alpha1.ListenerReasonInvalidRoutesRef),
+					Message:            msg,
+				})
+
+				logger.Errorf(msg)
+				listenerInError = true
+				continue
+			}
+
+			// TODO: support RouteNamespaces
+			httpRoutes, err := client.GetHTTPRoutes(gateway.Namespace, labels.SelectorFromSet(listener.Routes.Selector.MatchLabels))
+			if err != nil {
+				msg := fmt.Sprintf("Cannot fetch HTTPRoutes for namespace %q and matchLabels %v", gateway.Namespace, listener.Routes.Selector.MatchLabels)
+				// update "ResolvedRefs" status true with "InvalidRoutesRef" reason
+				listenerStatuses[i].Conditions = append(listenerStatuses[i].Conditions, metav1.Condition{
+					Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionFalse,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(v1alpha1.ListenerReasonInvalidRoutesRef),
+					Message:            msg,
+				})
+
+				logger.Errorf(msg)
+				listenerInError = true
+				continue
+			}
+
+			for _, httpRoute := range httpRoutes {
+				// Should never happen
+				if httpRoute == nil {
```

Comment:
Are you sure that this check is really necessary? 

I think the Kubernetes provider will not return an array with `nil` inside.